### PR TITLE
host: dai: add support for variable data copy

### PR DESCRIPTION
**Pull request by tlauda**
Adds support to host and dai components for variable data size copy.

### CODE REVIEW

File: src/drivers/dw/dma.c [hunk 1]
```diff
@@ -526,8 +530,15 @@ static int dw_dma_release(struct dma *dma, int channel)
 	spin_lock_irq(&dma->lock, flags);
 
 	/* get next lli for proper release */
-	p->chan[channel].lli_current =
-		(struct dw_lli2 *)p->chan[channel].lli_current->llp;
+	chan->lli_current = (struct dw_lli2 *)chan->lli_current->llp;
+
+	/* copy leftover data between current and last lli */
+	start_ptr = DW_DMA_LLI_ADDRESS(chan->lli_current, chan->direction);
+	bytes_left = start_ptr - (int32_t)chan->ptr_data.current_ptr;
```
Comment (lgirdwood) [inline src/drivers/dw/dma.c]: What happens if current ptr has bit31 set here (as we are casting to signed) ? start ptr should always be <= current ptr.

Comment (tlauda) [reply src/drivers/dw/dma.c]: @lgirdwood Good catch. I've fixed it by doing manual comparison on uints and also changed start_ptr to next_ptr as it's better shows the intentions here.

**Comment by ranj063** [comment]
@tlauda for the sake of people like me who are not an expert in DMA, could you please please add an explanation in the commit message saying what variable data copy is really about?

**Comment by tlauda** [comment]
@ranj063 Right now we are assuming that every pipeline_copy we have at least 1 period of data. It's maybe true for the DMA IRQ based scheduling, but for other types of scheduling like e.g. timer based it isn't exactly valid assumption. I've managed to refactor the code in such way, that now it always copy as much data as it can without any limitations. It simply reads read/write pointers from the DMAs and uses it to calculate the current data size. And I've managed to do it transparently to the scheduler.

**Comment by xiulipan** [comment]
@tlauda 
I checked the fail of APL_UP2_PCM512X platform
it seems the error is because of some random codec error?
```
[   88.560038 <    0.012985>] pcm512x i2c-104C5122:00: No SCLK, using BCLK: -2
[   88.562008 <    0.001970>] sof-audio-pci 0000:00:0e.0: pcm: open stream 0 dir 0
[   88.562013 <    0.000005>] sof-audio-pci 0000:00:0e.0: period min 192 max 16384 bytes
[   88.562015 <    0.000002>] sof-audio-pci 0000:00:0e.0: period count 2 max 16
[   88.562017 <    0.000002>] sof-audio-pci 0000:00:0e.0: buffer max 65536 bytes
```
But For BYT, all other tests are broken. Not sure what happened. Need to check.

**Comment by tlauda** [comment]
@lgirdwood I think it's ready to be merged. I'm not sure why tools build is stuck, some kind of CI environment problem.

**Comment by lgirdwood** [comment]
@tlauda yeah, 2 hours later it is still "pending", but travis build passed.

**Pull request merged**